File: FastStyleTransfer/FastStyleTransferDemo.cs
using AlbiruniML;
using alb = AlbiruniML.Ops;
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Drawing.Drawing2D;
using System.Drawing.Imaging;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Windows.Forms;
using System.Diagnostics;
using System.Threading;

namespace FastStyleTransfer
{
    public partial class FastStyleTransferDemo : Form
    {
        public FastStyleTransferDemo()
        {
            InitializeComponent();
        }
        TransformNet tnet;
        public static Bitmap ResizeImage(Image image, int width, int height)
        {
            var destRect = new Rectangle(0, 0, width, height);
            var destImage = new Bitmap(width, height);

            destImage.SetResolution(image.HorizontalResolution, image.VerticalResolution);

            using (var graphics = Graphics.FromImage(destImage))
            {
                graphics.CompositingMode = CompositingMode.SourceCopy;
                graphics.CompositingQuality = CompositingQuality.HighQuality;
                graphics.InterpolationMode = InterpolationMode.HighQualityBilinear;
                graphics.SmoothingMode = SmoothingMode.HighQuality;
                graphics.PixelOffsetMode = PixelOffsetMode.HighQuality;

                using (var wrapMode = new ImageAttributes())
                {
                    wrapMode.SetWrapMode(WrapMode.TileFlipXY);
                    graphics.DrawImage(image, destRect, 0, 0, image.Width, image.Height, GraphicsUnit.Pixel, wrapMode);
                }
            }

            return destImage;
        }
        Tensor imageTensor;
        private Tensor LoadImage(OpenFileDialog ofd)
        {
            Image image = ResizeImage(Bitmap.FromFile(ofd.FileName), 500, 500);
            SrcPictureBox.Image = image;
            var x = alb.buffer(new int[] { image.Height, image.Width, 3 });
            Bitmap bmp = new Bitmap(image);
            for (int i = 0; i < bmp.Height; i++)
            {
                for (int j = 0; j < bmp.Width; j++)
                {
                    Color clr = bmp.GetPixel(j, i);
                    float red = clr.R / 255.0f;
                    float green = clr.G / 255.0f;
                    float blue = clr.B / 255.0f;

                    x.Set(red, i, j, 0);
                    x.Set(green, i, j, 1);
                    x.Set(blue, i, j, 2);
                }
            }
            return x.toTensor();
        }

        private void LoadTensor(object dataobject)
        {
            Tensor data = dataobject as Tensor;
            Bitmap bmp = new Bitmap(data.Shape[0], data.Shape[1]);
            for (int i = 0; i < bmp.Height; i++)
            {
                for (int j = 0; j < bmp.Width; j++)
                {
                    var col = Color.FromArgb((int)(data.Get(i, j, 0) * 255.0f),
                        (int)(data.Get(i, j, 1) * 255.0f),
                        (int)(data.Get(i, j, 2) * 255.0f));
                    bmp.SetPixel(j, i, col);

                }
            }
            DisPictureBox.Image = bmp;
            this.label1.Visible = false;

            this.progressBar1.Visible = false;
            this.SaveButton.Enabled = true;
            this.TransferButton.Enabled = true;
            this.StyleComboBox.Enabled = true;
            this.BrowseButton.Enabled = true;
        }

        private void BrowseButton_Click(object sender, EventArgs e)
        {
            OpenFileDialog ofd = new OpenFileDialog();
            var resp = ofd.ShowDialog();
            if (resp == System.Windows.Forms.DialogResult.OK)
            {
                this.imageTensor = LoadImage(ofd);

            }
            this.TransferButton.Enabled = true;
        }
        delegate void FormInvok();
        delegate void FormInvokWithParam(object param);
        Thread t;
        private void TransferButton_Click(object sender, EventArgs e)
        {
            this.TransferButton.Enabled = false;
            this.label1.Visible = true;

            this.progressBar1.Visible = true;
            this.progressBar1.Value = 0;
            this.progressBar1.Maximum = 120;
            this.StyleComboBox.Enabled = false;
            this.BrowseButton.Enabled = false;
            t = new Thread(new ThreadStart(Transfer));
            t.Start();
            
        }

        private void Transfer()
        {
            this.imageTensor.keep();

            var p = tnet.Predict(this.imageTensor);
            this.Invoke(new FormInvokWithParam(LoadTensor), p); 
        }

        private void FastStyleTransferDemo_Load(object sender, EventArgs e)
        {
            tnet = new TransformNet();
            tnet.ReportProgress += tnet_ReportProgress;
            foreach (var item in tnet.variableDictionary)
            {
                this.StyleComboBox.Items.Add(item.Key);
            }

            this.StyleComboBox.SelectedItem = this.StyleComboBox.Items[0];
            this.StylePictureBox.Image = new Bitmap("styles/" + this.StyleComboBox.Items[0].ToString() + ".jpg");
        }

        void tnet_ReportProgress(int progress)
        {
            this.Invoke(new FormInvokWithParam(UpdateProgress), progress);
        }
        void UpdateProgress(object value)
        {
            this.progressBar1.Value = Convert.ToInt32(value);
        }
        private void StyleComboBox_SelectedIndexChanged(object sender, EventArgs e)
        {
            this.StylePictureBox.Image = new Bitmap("styles/" + this.StyleComboBox.SelectedItem.ToString() + ".jpg");
            this.tnet.ChangeVariable(this.StyleComboBox.SelectedItem.ToString());
           
        }

        private void SaveButton_Click(object sender, EventArgs e)
        {
            SaveFileDialog sfd = new SaveFileDialog();
            var sd = sfd.ShowDialog();
            if (sd == System.Windows.Forms.DialogResult.OK)
            {

                DisPictureBox.Image.Save(sfd.FileName);
            }
        }

        private void FastStyleTransferDemo_FormClosing(object sender, FormClosingEventArgs e)
        {
            if (t != null)
            {
                if (t.IsAlive)
                {
                    t.Abort();
                }
            }
        }
   
    }
}

File: FastStyleTransfer/FastStyleTransferDemo.vb
Imports AlbiruniML
Imports alb = AlbiruniML.Ops
Imports System
Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.Data
Imports System.Drawing
Imports System.Drawing.Drawing2D
Imports System.Drawing.Imaging
Imports System.Linq
Imports System.Text
Imports System.Threading.Tasks
Imports System.Windows.Forms
Imports System.Diagnostics
Imports System.Threading

Namespace FastStyleTransfer
    Public Partial Class FastStyleTransferDemo
        Inherits Form

        Public Sub New()
            InitializeComponent()
        End Sub

        Private tnet As TransformNet

        Public Shared Function ResizeImage(ByVal image As Image, ByVal width As Integer, ByVal height As Integer) As Bitmap
            Dim destRect = New Rectangle(0, 0, width, height)
            Dim destImage = New Bitmap(width, height)
            destImage.SetResolution(image.HorizontalResolution, image.VerticalResolution)

            Using graphics = Graphics.FromImage(destImage)
                graphics.CompositingMode = CompositingMode.SourceCopy
                graphics.CompositingQuality = CompositingQuality.HighQuality
                graphics.InterpolationMode = InterpolationMode.HighQualityBilinear
                graphics.SmoothingMode = SmoothingMode.HighQuality
                graphics.PixelOffsetMode = PixelOffsetMode.HighQuality

                Using wrapMode = New ImageAttributes()
                    wrapMode.SetWrapMode(WrapMode.TileFlipXY)
                    graphics.DrawImage(image, destRect, 0, 0, image.Width, image.Height, GraphicsUnit.Pixel, wrapMode)
                End Using
            End Using

            Return destImage
        End Function

        Private imageTensor As Tensor

        Private Function LoadImage(ByVal ofd As OpenFileDialog) As Tensor
            Dim image As Image = ResizeImage(Bitmap.FromFile(ofd.FileName), 500, 500)
            SrcPictureBox.Image = image
            Dim x = alb.buffer(New Integer() {image.Height, image.Width, 3})
            Dim bmp As Bitmap = New Bitmap(image)

            For i As Integer = 0 To bmp.Height - 1

                For j As Integer = 0 To bmp.Width - 1
                    Dim clr As Color = bmp.GetPixel(j, i)
                    Dim red As Single = clr.R / 255.0F
                    Dim green As Single = clr.G / 255.0F
                    Dim blue As Single = clr.B / 255.0F
                    x.[Set](red, i, j, 0)
                    x.[Set](green, i, j, 1)
                    x.[Set](blue, i, j, 2)
                Next
            Next

            Return x.toTensor()
        End Function

        Private Sub LoadTensor(ByVal dataobject As Object)
            Dim data As Tensor = TryCast(dataobject, Tensor)
            Dim bmp As Bitmap = New Bitmap(data.Shape(0), data.Shape(1))

            For i As Integer = 0 To bmp.Height - 1

                For j As Integer = 0 To bmp.Width - 1
                    Dim col = Color.FromArgb(CInt((data.[Get](i, j, 0) * 255.0F)), CInt((data.[Get](i, j, 1) * 255.0F)), CInt((data.[Get](i, j, 2) * 255.0F)))
                    bmp.SetPixel(j, i, col)
                Next
            Next

            DisPictureBox.Image = bmp
            Me.label1.Visible = False
            Me.progressBar1.Visible = False
            Me.SaveButton.Enabled = True
            Me.TransferButton.Enabled = True
            Me.StyleComboBox.Enabled = True
            Me.BrowseButton.Enabled = True
        End Sub

        Private Sub BrowseButton_Click(ByVal sender As Object, ByVal e As EventArgs)
            Dim ofd As OpenFileDialog = New OpenFileDialog()
            Dim resp = ofd.ShowDialog()

            If resp = System.Windows.Forms.DialogResult.OK Then
                Me.imageTensor = LoadImage(ofd)
            End If

            Me.TransferButton.Enabled = True
        End Sub

        Friend Delegate Sub FormInvok()
        Friend Delegate Sub FormInvokWithParam(ByVal param As Object)
        Private t As Thread

        Private Sub TransferButton_Click(ByVal sender As Object, ByVal e As EventArgs)
            Me.TransferButton.Enabled = False
            Me.label1.Visible = True
            Me.progressBar1.Visible = True
            Me.progressBar1.Value = 0
            Me.progressBar1.Maximum = 120
            Me.StyleComboBox.Enabled = False
            Me.BrowseButton.Enabled = False
            t = New Thread(New ThreadStart(AddressOf Transfer))
            t.Start()
        End Sub

        Private Sub Transfer()
            Dim sw As Stopwatch = New Stopwatch()
            sw.Start()
            Me.imageTensor.keep()
            Dim p = tnet.Predict(Me.imageTensor)
            Me(New FormInvokWithParam(AddressOf LoadTensor), p)
            sw.[Stop]()
            MessageBox.Show(sw.ElapsedMilliseconds.ToString())
        End Sub

        Private Sub FastStyleTransferDemo_Load(ByVal sender As Object, ByVal e As EventArgs)
            tnet = New TransformNet()
            AddHandler tnet.ReportProgress, AddressOf tnet_ReportProgress

            For Each item In tnet.variableDictionary
                Me.StyleComboBox.Items.Add(item.Key)
            Next

            Me.StyleComboBox.SelectedItem = Me.StyleComboBox.Items(0)
            Me.StylePictureBox.Image = New Bitmap("styles/" & Me.StyleComboBox.Items(0).ToString() & ".jpg")
        End Sub

        Private Sub tnet_ReportProgress(ByVal progress As Integer)
            Me(New FormInvokWithParam(AddressOf UpdateProgress), progress)
        End Sub

        Private Sub UpdateProgress(ByVal value As Object)
            Me.progressBar1.Value = Convert.ToInt32(value)
        End Sub

        Private Sub StyleComboBox_SelectedIndexChanged(ByVal sender As Object, ByVal e As EventArgs)
            Me.StylePictureBox.Image = New Bitmap("styles/" & Me.StyleComboBox.SelectedItem.ToString() & ".jpg")
            Me.tnet.ChangeVariable(Me.StyleComboBox.SelectedItem.ToString())
        End Sub

        Private Sub SaveButton_Click(ByVal sender As Object, ByVal e As EventArgs)
            Dim sfd As SaveFileDialog = New SaveFileDialog()
            Dim sd = sfd.ShowDialog()

            If sd = System.Windows.Forms.DialogResult.OK Then
                DisPictureBox.Image.Save(sfd.FileName)
            End If
        End Sub

        Private Sub FastStyleTransferDemo_FormClosing(ByVal sender As Object, ByVal e As FormClosingEventArgs)
            If t IsNot Nothing Then

                If t.IsAlive Then
                    t.Abort()
                End If
            End If
        End Sub
    End Class
End Namespace
